### ISSUE TITLE
Typescript: Support go-to-definition in diff editor

### ISSUE THREAD

**Issue by josh-tepper** [Microsoft/vscode#34034]
- VSCode Version: Code 1.16.0 (787b31c0474e6165390b5a5989c9619e3e16f953, 2017-09-06T16:12:42.401Z)
- OS Version: Darwin x64 15.6.0
- Extensions: 

Extension|Author (truncated)|Version
---|---|---
vscode-docker|Pet|0.0.16
ActiveFileInStatusBar|Ros|1.0.3
gitignore|cod|0.5.0
githistory|don|0.2.3
python|don|0.7.0
gitlens|eam|4.4.3
jsonnet|hep|0.0.13
Go|luk|0.6.65
cpptools|ms-|0.12.3
Ruby|reb|0.15.0
java|red|0.9.0
vscode-hexdump|sle|1.4.0
vscode-open-in-github|sys|1.6.0
vscode-proto3|zxh|0.1.2

---
I believe this regression happened about 2 versions ago:
It used to be the case that I could go to definitions directly from the right hand pane of the diff editor (ie, from the pane that represents the _current_ version of the file).  This is no longer the case. This seems to affect both languages that I regularly use: Java and Go

Reproduces without extensions: Yes/No
N/A -- with disable-extensions, go to definition is not available.

**Comment by roblourens** [comment]
This still works in TS/JS, so please file an issue on the Java or Go extensions.

**Comment by josh-tepper** [comment]
@roblourens -- I did some exploration -- this *does* work for both java/go for unstaged changes, but for neither with changes that are already in the index.  Thus I think that this is a vscode issue.  Can you re-open?

**Comment by ramya-rao-a** [comment]
I can confirm @josh-tepper's findings

Also, in the diff for working files, when you right click, you get the below commands in the context menu
![image](https://user-images.githubusercontent.com/16890566/30349144-c74b84a8-97c6-11e7-96e4-1518aef68c82.png)


In the diff for staged files, you get only the `Copy` command

**Comment by alexandrudima** [comment]
When used with git, the diff editor can be used to compare three different states:
1. when clicking on a file that is staged (in the git viewlet), the left hand side shows the HEAD file version in the left compared with the index file version in the right. It is possible that neither file "exists" on disk. AFAIK TS, JS nor Go should work on these as the right hand side does not necessarily exist on disk. **Also, the right hand side is readonly in this case**.

2. when clicking on a file that is not staged (in the git viewlet), the left hand side shows the index file version in the left compared with the file on disk file version in the right. AFAIK TS, JS and Go should work on these as the right hand side is a file on disk. **The right hand side is editable in this case**.

There is little in terms of UI except for the title of the diff editor to distinguish these two cases. AFAIK nothing has changed here, but perhaps @joaomoreno has some ideas / can double check my description is correct ?

@josh-tepper Was it the case that in case 1. above Java and Go would provide cross-file functionalities? Because that seems like a bug of Java and Go, how could they build a language server on the state contained in the git index...

**Comment by josh-tepper** [comment]
I'm not sure if they used to provide this functionality or I just never noticed the distinction before.  However:

In the cases that frustrated me, the index version was the one on disk.  Consider the following two workflows (both relevant/frequent for me):
- One is preparing a commit, stages the relevant files in the index, then visually confirms the change by clicking through the diff views of the staged files.  It is still possible (and indeed it has happened often) that I've wanted to follow a reference in one of these diff views, and in these cases, the index file is the one on disk!  It makes no difference to me that it is read only.
- I want to see an entire topic branche's diff (perhaps as reviewing a PR in github), so I check out the branch, then `git reset --soft $(git mergebase origin/master @)` so that I can see the entire diff in the source control view (the entire diff becomes "staged" on top of the base commit).  In this case, it is also true that the index version is the version on disk, and it is again, not relevant that it is read only.

In any case, two thoughts on improving vscodes behavior from an "outsider":
- It is possible to determine if there are any changes between the index and the working tree (if there are, these changes are listed in the source control view, so vscode seems to be aware of them).  Perhaps the behavior could depend on this.
- Regardless as to whether or not the behavior depends on this, it seems like the error message could be changed to something like "not available on index files" or if the behavior is changed as suggested "not available for files that differ from the working tree".

Let me know what you guys think

**Comment by josh-tepper** [comment]
I can see how this might be very messy to implement though.  The working tree diffs seem to come from git, and they are sometimes stale prior to being refreshed.  Further, a file can change at any moment.  If you reference the on disk file, but the on disk file actually differs from the index, the behavior would be very strange.

At the very least a better error message would help a lot.  I really wish though that there was a better way to address this situation more completely, at least in limited circumstances.

**Comment by joaomoreno** [comment]
This really needs to come from TypeScript. The right hand side of an index change is a read-only, staged version of the file, not necessarily equal to the contents in disk.

It would require TypeScript to kind of _fork_ the language server at some point in time and replace the contents of the file in disk with the contents of the staged file in order to get language features on it.

@josh-tepper How do you trigger the action exactly?

**Comment by josh-tepper** [comment]
With the keyboard command for editor.action.goToDeclaration (default f12, f3 in my world)

**Comment by josh-tepper** [comment]
@joaomoreno  @ramya-rao-a -- I actually tend to disagree that the language plugin should handle this.  From my (admittedly fuzzy) understanding, the language server protocol doesn't include the concept of multiple copies/versions of the same files.  Further, as discussed in the issue against the java plugin (see: https://github.com/redhat-developer/vscode-java/issues/295), the string that the plugin gets looks like this:
```
git:/Users/fbricon/Dev/projects/vscode-java-demo/src/main/java/com/redhat/vscode/demo/DevoxxApplication.java?%7B%22path%22%3A%22%2FUsers%2Ffbricon%2FDev%2Fprojects%2Fvscode-java-demo%2Fsrc%2Fmain%2Fjava%2Fcom%2Fredhat%2Fvscode%2Fdemo%2FDevoxxApplication.java%22%2C%22ref%22%3A%22%22%7D
```

Thus, it seems like the specifics of the way this is handled depends on the source control in use.  I'm not sure that these plugins should have to know anything about source control or how to parse a string like the one above.

On the other hand, if this is handled by vscode itself, it can choose to only ever send "working tree" paths to the plugins, which seems to fit my understanding of their expectations.  Further, vscode can produce a meaningful error if there is no equivalent file in the working tree to send.

It should be easy to determine if their is an equivalent working tree file with git, e.g., by taking the git hash of the blob for the index file and comparing with `git hash-obect -w <filename>`, or perhaps more simply, by checking whether the file has unstaged changes via the usual mechanisms.  If this check is initiated any time that the diff window is loaded or modified, you can be sure that the version from the diff is the one that was checked.

Let me know what you think

**Comment by joaomoreno** [comment]
> Thus, it seems like the specifics of the way this is handled depends on the source control in use.

Not really. This is independent from source control. It's really a matter of taking the language server state at a certain point, snapshot it and replacing one file's contents with something else. Then, the language server could actually answer interesting questions about those contents.

cc @jrieken for considering some UX when running F12 on a buffer location where that's not supported

**Comment by josh-tepper** [comment]
Ok.  But then is it not the case that each plugin would need to implement this separately?

**Comment by ramya-rao-a** [comment]
The Go extension registered all the providers only for the file scheme and so the definition provider never got called in the first place as the file in the staged diff view used the git scheme.

Fixed that, now the context menu shows the commands, the definition provider gets called correctly, the document passed corresponds to the one in the index which is used to find the definition

Closing this issue as the work needs to be done on the extension's side.

Thanks @joaomoreno and @alexandrudima for the clarifications

**Comment by alexandrudima** [comment]
@ramya-rao-a So is the plan that you bake git knowledge into all the Go tools? How can the Go tools go to definition across files that live only in the git index and are not on disk ?

Consider the following:

(INDEX):
 * file1.go contains Go code
 * file2.go contains Go code

(FILE SYSTEM):
 * file1.go is empty
 * file2.go is empty

The diff editor is brought by comparing (HEAD) with (INDEX). The right-hand side of the diff editor is read only and shows the (INDEX) version of file2.go.

How could the go tools compute anything across files unless they all bake in git knowledge ?

**Comment by ramya-rao-a** [comment]
@alexandrudima 

Let's take the case of `Go to Definition` feature.
The Go tool to find the information needs
- file contents with position so that it can determine the symbol for which definition needs to be found
     - this is not a problem now that I have removed the constraint on the file scheme. The document as per INDEX is passed to the definition provider
- The state of the complete file system as per INDEX. 
     - This is something that we (the Go extension) cannot do and I dont intend to do it either. A language server may attempt to do this, but currently the Go extension is built using a set of independent tools that work off of the files in the file system

So my plan is
- Enable the Go to/Peek definition, hover info, find all references features to show up in the context menu in the diff view
- The above will use the state of the files in the file system to find definition/references

**Comment by alexandrudima** [comment]
@ramya-rao-a 

---

when diffing (HEAD)<->(INDEX) (i.e. modified file appears in STAGED CHANGES):

![image](https://user-images.githubusercontent.com/5047891/30704988-88e7726e-9ef4-11e7-9b01-6b641068cf32.png)

the left hand side is:
**A** - `git:/Users/alex/src/working_dir/.vscode/settings.json?%7B%22path%22%3A%22%2FUsers%2Falex%2Fsrc%2Fworking_dir%2F.vscode%2Fsettings.json%22%2C%22ref%22%3A%22HEAD%22%7D`
the right hand side is:
**B** - `git:/Users/alex/src/working_dir/.vscode/settings.json?%7B%22path%22%3A%22%2FUsers%2Falex%2Fsrc%2Fworking_dir%2F.vscode%2Fsettings.json%22%2C%22ref%22%3A%22%22%7D`

---

when diffing (INDEX)<->(WORKING TREE) (i.e. modified file appears in CHANGES):

![image](https://user-images.githubusercontent.com/5047891/30705047-ace67c8c-9ef4-11e7-980f-3aeec3765617.png)

the left hand side is:
**B** - `git:/Users/alex/src/working_dir/.vscode/settings.json?%7B%22path%22%3A%22%2FUsers%2Falex%2Fsrc%2Fworking_dir%2F.vscode%2Fsettings.json%22%2C%22ref%22%3A%22~%22%7D`
the right hand side is:
**C** - ` file:///Users/alex/src/working_dir/.vscode/settings.json`

---

if you register things for `file://` URIs the extension will work correctly. It is incorrect to register cross-file providers for `git://` URIs unless your language service knows how to do git. i.e. load all the files at the state of the (INDEX) or at the state of the (HEAD).

Your statement is very confusing, as it appears the Go extension was correct before... :
> The Go extension registered all the providers only for the file scheme and so the definition provider never got called in the first place as the file in the staged diff view used the git scheme.

> Fixed that, now the context menu shows the commands, the definition provider gets called correctly, the document passed corresponds to the one in the index which is used to find the definition

**Comment by ramya-rao-a** [comment]
> it is not possible to go to definition from git://index/fileB.go to git://index/fileA.go unless the go tools know how to read from the git object database...

That's right. And the Go tools do not know how to read from the git object database. Thats not going to change.

> if a symbol is defined in fileA.go (which has been deleted or emptied on disk).
and that symbol is used in `git://index/fileB.go` (but not in `fileB` on disk)
it is not possible to go to definition from `git://index/fileB.go` to `git://index/fileA.go`

True. I am not promising a feature where one can find definition/references among staged files only. I am only enabling the find definition/references features to be run from the staged diff view. They will operate on files on disk.

I can add a notification in such cases warning the user that the extension will not be looking at staged files, but will look at files on disk when attempting to find definitions/references to make it clear

**Comment by alexandrudima** [comment]
@ramya-rao-a I am sorry, perhaps I am explaining this wrong. @mjbvz This is also a problem for TypeScript it appears:

Here are the steps for a cross-file language like TypeScript:
* in an empty folder, initialize a git repository
* create an empty file `fileA.ts` and another empty file `fileB.ts` and stage both files.
* commit both empty files
* at this point, the repo contains two empty files and git status shows nothing.
---
* open `fileA.ts`, edit it and save it:
```ts
export class Provider {
    public provide(): string {
        return 'hello world!';
    }
}
```
* open `fileB.ts`, edit it and save it:
```ts
import { Provider } from './fileA';
let p: Provider;
p.provide();
```
* stage both `fileA.ts` and `fileB.ts`
* open `fileA.ts`. empty its content and save it
* open `fileB.ts`. empty its content and save it

---

The git viewlet should look something like this:
![image](https://user-images.githubusercontent.com/5047891/30734860-363c74a4-9f7d-11e7-8005-3fe4fd8d47c2.png)

---

Git status should show something like this:
![image](https://user-images.githubusercontent.com/5047891/30734877-49c8e19c-9f7d-11e7-8c39-cf6beb6d8e76.png)

---

Click on `fileB.ts` in the "STAGED CHANGES"

Observe how **incorrectly** Go to Definition is available when right clicking. This is wrong. I was hoping to use TS as a "this is the correct way", but TS is also wrong. cc @mjbvz 

![image](https://user-images.githubusercontent.com/5047891/30734950-9220e5e8-9f7d-11e7-8435-5e77f1ea0027.png)

---

Note how this can never go to definition correctly unless TS begins to read files in the git index. The same logic applies to Go.

TL;DR: IMHO cross-file language features should not be registered for `git://` URIs unless the language server knows how to read files from the git object database / index.

I have attached the git repository as it was on my machine if you have trouble reproducing:

[working_dir3.zip](https://github.com/Microsoft/vscode/files/1323839/working_dir3.zip)

---

cc @dbaeumer @joaomoreno if they have a better way to explain the conceptual problem than I am.

**Comment by ramya-rao-a** [comment]
@alexandrudima I understood the conceptual problem the first time you tried to explain it back in https://github.com/Microsoft/vscode/issues/34034#issuecomment-330857328 :)

If our stance is that `Cross file features when operated on staged files in the diff view should always be accurate`, then you are right: The features/providers shouldnt register for `git://` URIs unless the language server knows how to read files from the git object database /index .

My stance in the purview of the Go extensions was `Cross file features when operated on staged files in the diff view will try their best using on disk files.` This will not be always accurate as the files on disk can differ from files on Index`. This will be accurate when there are no unstaged changes along with the staged changes.

There is no way the Go extension will work off of the git index unless it moves to a language server model. So this was me trying to make the cross file features available, knowing fully well that they will not be accurate when there are both staged and unstaged changes. I can warn the user of the same.

**Comment by dbaeumer** [comment]
IMO we need to somehow let the user know that these results are inaccurate. Otherwise we will get new issues complaining about inaccurate results which we will very likely never address.

**Comment by alexandrudima** [comment]
Thank you @ramya-rao-a for confirming you understand my concern.

Indeed, I hold the belief that semantic language features **should always be accurate**. My motivation is *trust* related. I believe there is an implicit trust relationship between a user and a tool, and with each incorrectness, where the tool disappoints the user, the trust relationship slowly erodes, to the point the user dumps the tool for another one, which disappoints less.

**Comment by ramya-rao-a** [comment]
@alexandrudima Point taken. https://github.com/Microsoft/vscode-go/commit/b63ebb8ab901ad17c48ff04a0c2f416f8048f118
